ksql-engine/src/test/resources/query-validation-tests/README.md

```diff
@@ -5,9 +5,191 @@ file contains one or more tests. They are written in a simple json format that d
-for joins etc, but currently only the final topology will be verified. This should be enough,
+The test cases are run by the `QueryTranslationTest` test class.
+
+## Topology comparision
+These test also validate the generated topology matches the expected topology,
+i.e. a test will fail if the topology has changed from previouus runs.
+This is needed to detect potentially non-backwards compatible changes to the generated topology.
+
+The expected topology files, and the configuration used to generated them are found in
+`src/test/resources/expected_topology/<Version Number>`
+
+By default test will compare the current generated topology against the previous released version
+of KSQL, as identified by the `QueryTranslationTest.CURRENT_TOPOLOGY_VERSION` variable.
+
+To run this test against previously released versions there are three options
+
+1. Manually change `QueryTranslationTest.CURRENT_TOPOLOGY_VERSION` to a valid version number
+found under the `src/test/resources/expected_topology` directory.
+
+1. Set the system property "topology.version" to the version required.
+  * Within Intellij
+    1. Click Run/Edit configurations
+    1. Select the QueryTranslationTest
+    1. Enter `-Dtopology.version=X` in the "VM options:" form entry
+       where X is the desired previously released version number.
+  * From the command line
+    1. run `mvn clean package -DskipTests=true` from the base of the KSQL project
+    1. Then run `mvn test -Dtopology.version=X -Dtest=QueryTranslationTest -pl ksql-engine`.
+       Again X is the version you want to run the tests against.
+
+  Note that for both options above the version must exist
+  under the `src/test/resources/expected_topology` directory.
+
+For instructions on how to generate new topologies, see `TopologyFileGenerator.java`
+
+## Adding new tests
+
+The following is a template test file:
+
+```json
+{
+  "comments": [
+    "Add a description of _what_ are of functionality this file is testing"
+  ],
+  "tests": [
+    {
+      "name": "my first positive test",
+      "description": "an example positive test where the output is verified"
+      "statements": [
+        "CREATE STREAM intput (ID bigint) WITH (kafka_topic='input_topic', value_format='JSON');",
+        "CREATE STREAM output AS SELECT id FROM test WHERE id < 10;"
+      ],
+      "inputs": [
+        {"topic": "input_topic", "key": 0, "value": "{"id": 8}", "timestamp": 0},
+        {"topic": "input_topic", "key": 0, "value": "{"id": 10}", "timestamp": 10000},
+        {"topic": "input_topic", "key": 1, "value": "{"id": 9}", "timestamp": 30000},
+        {"topic": "input_topic", "key": 1, "value": "{"id": 11}", "timestamp": 40000}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": 0, "value": "{"id": 8}", "timestamp": 0},
+        {"topic": "OUTPUT", "key": 0, "value": "{"id": 9}", "timestamp": 30000}
+     ]
+    },
+    {
+      "name": "my first negative test",
```

Comment:
tbh, I'm not sure I agree that this is the right place for negative tests - it seems kind-of heavy to setup an end-to-end test just to test something that doesn't parse. I would be more inclined to create a simple parser unit test util that allows us to unit test negative examples

Reply:
I can see your point of view, but I feel that allowing negative tests in these JSON functional tests means we're more likely to see contributions testing that users are seeing sensible error messages for unsupported operations.  

I'm not particularly talking about the error a user will see if they mistype something. (I agree this belongs in a unit test). I'm talking more about statements that a user might think are valid, but aren't supported, e.g. GROUP BY functionality where the projection contains columns that aren't an exact match for the group by clause.  Such testing of what is _not_ supported, IMHO, belongs with the functional testing of what _is_ supported.